### PR TITLE
Gpcopy migration/using docs

### PR DESCRIPTION
**Pull request by dyozie**
This is the "using" document, based on the one for gptransfer.  HTML for this is being staged at http://docs-gpdb-review-staging.cfapps.io/review/admin_guide/managing/gpcopy-migrate.html

### CODE REVIEW

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    </topic>
+    <topic id="topic_nd3_2sp_zdb">
+    <title>Validating Copied Data</title>
+    <body>
+      <p>By default, <codeph>gpcopy</codeph> does not validate the data transferred. You can request
+        validation using the <codeph>--validate=<i>type</i></codeph> option. The validation
+          <i>type</i> can be one of the following:<ul id="ul_hf1_k21_xdb">
+          <li><codeph>count</codeph> - compares the row counts between the source and destination
+            tables.</li>
+          <li><codeph>md5xor</codeph> - calculates the MD5 value of all rows, then performs an XOR
```
Comment (Eulerizeit) [inline gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: If you are using this to go from 4->5 then the md5 won't match even if the tables are the same.

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            over the MD5 values.</li>
+        </ul></p>
+    </body>
+  </topic>
+    <topic id="topic_ytw_2sp_zdb">
+    <title>Addressing Failed Data Transfers</title>
+    <body>
+      <p>When <codeph>gpcopy</codeph> encounters errors and quits or is cancelled by the user, the
+        currently-running copy operations on tables in the destination database are rolled back.
+        Copy operations that have completed are not rolled back. </p>
```
Comment (Eulerizeit) [inline gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: What does this mean exactly? What is an operation in this context? If I'm copying 3 tables and 1.5 have been finished the the 1 stays the the other 2 don't? Can an operation be a subset of a complete table (or whatever the atomic unit is)?

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        command. </p>
+      <p>The <codeph>gpcopy</codeph> utility logs messages in log file
+            <codeph>gpcopy_<varname>date</varname>.log</codeph> the <codeph>~/gpAdminLogs</codeph>
+        directory on the master host. If you run multiple <codeph>gpcopy</codeph> commands on the
+        same day, the utility appends messages to that day's log file. </p>
+    </body>
+  </topic>
+  <topic id="topic_pyc_hpp_zdb">
+    <title>Migrating Data Between Clusters that Share Hardware</title>
+    <body>
```
Comment (Eulerizeit) [inline gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: I think the process needs to be spelled out in more detail.

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+          id="ul_g5k_22q_zdb">
+          <li>Back up all source data before using <codeph>gpcopy</codeph> with
+              <codeph>--truncate</codeph>.</li>
+          <li>Use the <codeph>--validate</codeph> option with <codeph>--truncate</codeph> to ensure
+            that data is successfully copied before source tables are truncated.</li>
+          <li>If you use <codeph>--truncate</codeph> to migrate data on shared hardware, also
+            include the <codeph>--no-compression</codeph> option to avoid the overhead cost of
+            compressing and uncompressing data that does not move across the network.</li>
+          <li>Consider migrating only portions of the your data at a time, so that you can fully
+            address table errors or validation failures that may occur during the copy
```
Comment (Eulerizeit) [inline gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: Also, it is important that the cluster be offline while doing this.

**Comment by BrianPivotal** [comment]
”The source and destination databases can be in the same Greenplum cluster or a different cluster“

In this version, gpcopy does not support copy in the same Greenplum cluster.

**Comment by BrianPivotal** [comment]
"gpcopy is installed with Pivotal Greenplum Database starting with versions 5.9.0 and 4.3.16.0."

I think should be 5.9.0 and 4.3.26.0. We will discuss in this week's shipit meeting.

**Comment by BrianPivotal** [comment]
”and have certificate-authenticated SSH access between all hosts in both systems.“

“SSH keys must be exchanged between the two systems before using gpcopy. When you install a Greenplum Database cluster, you set up all the master and segment hosts so that the Greenplum Database administrative user (gpadmin) can connect with SSH from every host in the cluster to any other host in the cluster without providing a password. The gpcopy utility requires this capability between every host in the source and destination clusters. First, ensure that the clusters have network connectivity with each other. Then, prepare a hosts file containing a list of all the hosts in both clusters, and use the gpssh-exkeys utility to exchange keys. See the reference for gpssh-exkeys in the .”

content related to ssh can be removed now. gpcopy does not require ssh. I think it also applies to the gpcopy doc.

**Comment by BrianPivotal** [comment]
For "Dependency" section, 
Add "Add dependency to pg_dump, pg_dumpall and psql tools. And also keep their version same as source database cluster." Normally, user runs gpcopy from source, so it will not be a problem. If user runs gpcopy from other servers like ETL server, they need to make sure that server has pg_dump, pg_dumpall and psql setup and has the same version as source.

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      database, or just selected tables. The source and destination databases can be in the same
+      Greenplum cluster or a different cluster, and the clusters can have different Greenplum
+      Database versions. For example, you can use <codeph>gpcopy</codeph> to migrate data from a
+      version 4.x Greenplum Database to version 5.x.</p>
+    <p>The <codeph>gpcopy</codeph> interface includes options to transfer one or more full
+      databases, or one or more database tables. A full database transfer includes the database
+      schema, table data, indexes, views, roles, user-defined functions, and resource queues. If a
+      copied table or database does not exist in the destination cluster, <codeph>gpcopy</codeph>
+      creates it automatically. The utility re-creates indexes on tables as necessary before copying
+      data.</p>
```
[the remark this answers is not in the document]
Comment (dyozie) [reply gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: That's mentioned in the development notes, but other instances describe re-creating the index after copying data.  @BrianPivotal ?

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml (hunk 1 above)
Comment (dyozie) [reply gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: Added a note saying to use rowcount validation in that situation.

File: gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml (hunk 3 above)
Comment (dyozie) [reply gpdb-doc/dita/admin_guide/managing/gpcopy-migrate.xml]: Agreed - hoping to get more info on how it all works.

**Comment by BrianPivotal** [comment]
For --truncate-source-after, most likely, use case is that user performs local gpcopy (either from 5.x - 5.x or 4.3 - 5.x, etc) and disk space is not enough (e.g. already more than 70% full). --truncate-source-after must be used together with --validate. For one table, after copy and validation, gpcopy performs truncate table at source to release disk space. We do not drop table because the main goal is to release disk space and also we need to consider performance impact.

**Comment by BrianPivotal** [comment]
"gpcopy is dependent on the pg_dump, pg_dumpall, and psql utilities installed with Greenplum Database. In most cases, you run gpcopy from a Greenplum Database cluster, so the dependencies are automatically met. If you run gpcopy on a remote server, such as an ETL system, you manually copy the gpcopy utility as well as pg_dump, pg_dumpall, and psql to that system."

since pg_dump, pg_dumpall have dependent libraries, it's better to use the corresponding version of Greenplum Clients (Greenplum Clients contains all the dependent libraries for pg_dump) together with gpcopy.

**Comment by Eulerizeit** [comment]
On Tue, Jun 5, 2018 at 1:25 AM, Brian Lu <notifications@github.com> wrote:

> "Note: If you are migrating data from Greenplum 4.x to 5.x, MD5 validation
> (--validate md5xor) will always fail. Use --validate count when migrating
> between major versions to validate by table row count."
>
> This can be removed. md5xor does not fail from 4.3 to 5.x.
>

Since the numeric data type changes on disk, how can the md5xor not fail?

-- Rob

**Comment by dyozie** [comment]
@BrianPivotal - I understand that gpcopy doesn't require `gpstart -R`.  But I think the intention is that if an admin is doing an actual migration (with validation) then they'd want to ensure that the source/destination table data isn't changing during the whole process.

**Comment by dyozie** [comment]
Merging this new content, as I believe all the feedback has been addressed.  We can still tweak things later if anything new is noticed.

**Pull request merged**